File: src/lib/api/flowOrchestrator.ts
import { invoke } from "@tauri-apps/api/core";
import type { InstallAuditSummary, InstallFileEntry } from "./designContract";

export interface WorkflowModule {
  id: string;
  name: string;
  nameZh?: string;
  source: string;
  description: string;
  capabilities: string[];
}

export interface WorkflowPresetSummary {
  id: string;
  name: string;
  nameZh?: string;
  description: string;
  r3Tier?: string;
  moduleCount: number;
  stageCount: number;
}

export interface WorkflowStage {
  id: string;
  name: string;
  prompt?: string;
  dependsOn?: string[];
  artifacts?: Array<{
    file: string;
    scope: string;
    optional?: boolean;
  }>;
}

export interface WorkflowPreset {
  id: string;
  name: string;
  nameZh?: string;
  description: string;
  r3Tier?: string;
  modules: string[];
  stages: WorkflowStage[];
  paths: Record<string, string[]>;
}

export interface FlowConfigGate {
  type: string;
  artifacts: string[];
}

export interface FlowConfigStage {
  id: string;
  enabled: boolean;
  depends_on?: string[];
  gates?: FlowConfigGate[];
}

export interface FlowConfigRules {
  max_auto_retry: number;
  role_separation: boolean;
  require_diff_boundary: boolean;
}

export interface FlowConfigReviewTier {
  enabled: boolean;
  lenses: string[];
  max_lenses?: number | null;
  default_lens?: string | null;
  paths?: string[];
  changed_lines_threshold?: number | null;
}

export interface FlowConfigReviewPolicy {
  mode: string;
  trivial: FlowConfigReviewTier;
  standard: FlowConfigReviewTier;
  hot_path: FlowConfigReviewTier;
  large_diff: FlowConfigReviewTier;
}

export interface FlowConfig {
  schema_version: number;
  preset_id: string;
  project_type: string;
  modules: string[];
  stages: FlowConfigStage[];
  rules: FlowConfigRules;
  review_policy: FlowConfigReviewPolicy;
  semantic_warnings?: string[];
}

export interface WorkflowProfile {
  schemaVersion: number;
  presetId: string;
  projectType: string;
  modules: string[];
  resolvedStages: string[];
  activeChangeId?: string | null;
  exportedAt: string;
  opensunstarVersion?: string;
  semanticWarnings?: string[];
}

export interface ArtifactStatus {
  file: string;
  relativePath: string;
  exists: boolean;
  nonEmpty: boolean;
  optional: boolean;
}

export interface TaskSummary {
  total: number;
  pending: number;
  inProgress: number;
  done: number;
  blocked: number;
}

export interface SpecsChangeIndex {
  changeId: string;
  artifactCompleteness: number;
  artifacts: ArtifactStatus[];
  taskSummary?: TaskSummary | null;
}

export interface SpecsWorkflowIndex {
  projectPath: string;
  workspaceExists: boolean;
  hasFlowKit: boolean;
  hasFlowConfig: boolean;
  hasSpecsDir: boolean;
  activeChangeId?: string | null;
  savedProfile?: WorkflowProfile | null;
  changes: SpecsChangeIndex[];
}

export interface StageGateResult {
  allowed: boolean;
  targetStage: string;
  changeId: string;
  missingArtifacts: string[];
  satisfiedArtifacts: string[];
  warnings: string[];
}

export interface FlowWritePlan {
  files: InstallFileEntry[];
  audit: InstallAuditSummary;
  semanticWarnings: string[];
}

export interface OrchestrationLogEntry {
  ts?: string | null;
  event: string;
  summary: string;
  payload: Record<string, unknown>;
}

export type OrchestrationStepStatus = "planned" | "applied" | "skipped" | "verified" | "failed";

export interface OrchestrationStepReceipt {
  id: string;
  label: string;
  targetPath: string;
  action: string;
  status: OrchestrationStepStatus;
  beforeChecksum?: string | null;
  afterChecksum?: string | null;
  snapshotPath?: string | null;
  reason?: string | null;
}

export interface OrchestrationVerification {
  id: string;
  label: string;
  passed: boolean;
  detail?: string | null;
}

export interface OrchestrationReceipt {
  schema: string;
  operation: string;
  projectPath: string;
  dryRun: boolean;
  createdAt: string;
  steps: OrchestrationStepReceipt[];
  verifications: OrchestrationVerification[];
}

export const flowOrchestratorApi = {
  async listModules(projectId?: string): Promise<WorkflowModule[]> {
    return await invoke<WorkflowModule[]>("list_workflow_modules_cmd", {
      projectId: projectId ?? null,
    });
  },

  async listPresets(projectId?: string): Promise<WorkflowPresetSummary[]> {
    return await invoke<WorkflowPresetSummary[]>("list_workflow_presets_cmd", {
      projectId: projectId ?? null,
    });
  },

  async getPreset(id: string, projectId?: string): Promise<WorkflowPreset> {
    return await invoke<WorkflowPreset>("get_workflow_preset_cmd", {
      id,
      projectId: projectId ?? null,
    });
  },

  async scanProject(
    projectId: string,
    presetId?: string,
    projectType?: string,
  ): Promise<SpecsWorkflowIndex> {
    return await invoke<SpecsWorkflowIndex>("scan_project_specs_workflow_cmd", {
      projectId,
      presetId: presetId ?? null,
      projectType: projectType ?? null,
    });
  },

  async validateStageGate(
    projectId: string,
    params: {
      presetId: string;
      projectType: string;
      changeId: string;
      targetStage: string;
    },
  ): Promise<StageGateResult> {
    return await invoke<StageGateResult>("validate_workflow_stage_gate_cmd", {
      projectId,
      ...params,
    });
  },

  async exportProfile(
    projectId: string,
    presetId: string,
    projectType: string,
    activeChangeId?: string,
    selectedModules?: string[],
    disabledStages?: string[],
    strictSemantics = true,
  ): Promise<WorkflowProfile> {
    return await invoke<WorkflowProfile>("export_project_workflow_profile_cmd", {
      projectId,
      presetId,
      projectType,
      activeChangeId: activeChangeId ?? null,
      selectedModules: selectedModules ?? null,
      disabledStages: disabledStages ?? null,
      strictSemantics,
    });
  },

  async previewProfileExport(
    projectId: string,
    presetId: string,
    projectType: string,
    activeChangeId?: string,
    selectedModules?: string[],
    disabledStages?: string[],
  ): Promise<FlowWritePlan> {
    return await invoke<FlowWritePlan>("preview_project_workflow_profile_export_cmd", {
      projectId,
      presetId,
      projectType,
      activeChangeId: activeChangeId ?? null,
      selectedModules: selectedModules ?? null,
      disabledStages: disabledStages ?? null,
    });
  },

  async exportFlowConfig(
    projectId: string,
    presetId: string,
    projectType: string,
    selectedModules?: string[],
    disabledStages?: string[],
    strictSemantics = true,
  ): Promise<FlowConfig> {
    return await invoke<FlowConfig>("export_flow_config_cmd", {
      projectId,
      presetId,
      projectType,
      selectedModules: selectedModules ?? null,
      disabledStages: disabledStages ?? null,
      strictSemantics,
    });
  },

  async previewFlowConfigExport(
    projectId: string,
    presetId: string,
    projectType: string,
    selectedModules?: string[],
    disabledStages?: string[],
  ): Promise<FlowWritePlan> {
    return await invoke<FlowWritePlan>("preview_flow_config_export_cmd", {
      projectId,
      presetId,
      projectType,
      selectedModules: selectedModules ?? null,
      disabledStages: disabledStages ?? null,
    });
  },

  async readOrchestrationLog(
    projectId: string,
    limit = 20,
  ): Promise<OrchestrationLogEntry[]> {
    return await invoke<OrchestrationLogEntry[]>("read_orchestration_log_cmd", {
      projectId,
      limit,
    });
  },

  async restoreLatestReceipt(projectId: string): Promise<OrchestrationReceipt> {
    return await invoke<OrchestrationReceipt>("restore_latest_orchestration_receipt_cmd", {
      projectId,
    });
  },
};

File: src-tauri/src/commands/flow_orchestrator.rs
//! SDD workflow orchestrator commands (flow-kit compatible indexing + gates).

use tauri::State;

use crate::database::Database;
use crate::services::flow_orchestrator::{
    export_flow_config, export_flow_config_strict, export_project_workflow_profile,
    export_project_workflow_profile_strict, get_workflow_preset, list_workflow_modules,
    list_workflow_presets, preview_flow_config_export, preview_project_workflow_profile_export,
    scan_project_specs_workflow, validate_workflow_stage_gate, FlowConfig, FlowWritePlan,
    SpecsWorkflowIndex, StageGateResult, WorkflowModule, WorkflowPreset, WorkflowPresetSummary,
    WorkflowProfile,
};
use crate::store::AppState;

fn project_path_for_id(db: &Database, project_id: &str) -> Result<String, String> {
    db.get_project(project_id)
        .map_err(|e| e.to_string())?
        .ok_or_else(|| format!("项目不存在: {project_id}"))
        .map(|p| p.path)
}

#[tauri::command]
pub async fn list_workflow_modules_cmd(
    state: State<'_, AppState>,
    project_id: Option<String>,
) -> Result<Vec<WorkflowModule>, String> {
    let project_path = match project_id {
        Some(id) => Some(project_path_for_id(&state.db, &id)?),
        None => None,
    };
    list_workflow_modules(project_path.as_deref()).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn list_workflow_presets_cmd(
    state: State<'_, AppState>,
    project_id: Option<String>,
) -> Result<Vec<WorkflowPresetSummary>, String> {
    let project_path = match project_id {
        Some(id) => Some(project_path_for_id(&state.db, &id)?),
        None => None,
    };
    list_workflow_presets(project_path.as_deref()).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn get_workflow_preset_cmd(
    state: State<'_, AppState>,
    id: String,
    project_id: Option<String>,
) -> Result<WorkflowPreset, String> {
    let project_path = match project_id {
        Some(pid) => Some(project_path_for_id(&state.db, &pid)?),
        None => None,
    };
    get_workflow_preset(&id, project_path.as_deref()).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn scan_project_specs_workflow_cmd(
    state: State<'_, AppState>,
    project_id: String,
    preset_id: Option<String>,
    project_type: Option<String>,
) -> Result<SpecsWorkflowIndex, String> {
    let path = project_path_for_id(&state.db, &project_id)?;
    scan_project_specs_workflow(&path, preset_id.as_deref(), project_type.as_deref())
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn validate_workflow_stage_gate_cmd(
    state: State<'_, AppState>,
    project_id: String,
    preset_id: String,
    project_type: String,
    change_id: String,
    target_stage: String,
) -> Result<StageGateResult, String> {
    let path = project_path_for_id(&state.db, &project_id)?;
    let result =
        validate_workflow_stage_gate(&path, &preset_id, &project_type, &change_id, &target_stage)
            .map_err(|e| e.to_string())?;

    if let Err(e) = crate::services::flow_orchestrator::append_orchestration_log(
        &path,
        serde_json::json!({
            "event": "stage_gate",
            "presetId": preset_id,
            "projectType": project_type,
            "changeId": change_id,
            "targetStage": target_stage,
            "allowed": result.allowed,
            "missing": result.missing_artifacts,
        }),
    ) {
        log::warn!("写入 orchestration log 失败: {e}");
    }

    Ok(result)
}

#[tauri::command]
pub async fn preview_project_workflow_profile_export_cmd(
    state: State<'_, AppState>,
    project_id: String,
    preset_id: String,
    project_type: String,
    active_change_id: Option<String>,
    selected_modules: Option<Vec<String>>,
    disabled_stages: Option<Vec<String>>,
) -> Result<FlowWritePlan, String> {
    let path = project_path_for_id(&state.db, &project_id)?;
    preview_project_workflow_profile_export(
        &path,
        &preset_id,
        &project_type,
        active_change_id.as_deref(),
        selected_modules.as_deref(),
        disabled_stages.as_deref(),
    )
    .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn export_project_workflow_profile_cmd(
    state: State<'_, AppState>,
    project_id: String,
    preset_id: String,
    project_type: String,
    active_change_id: Option<String>,
    selected_modules: Option<Vec<String>>,
    disabled_stages: Option<Vec<String>>,
    strict_semantics: Option<bool>,
) -> Result<WorkflowProfile, String> {
    let path = project_path_for_id(&state.db, &project_id)?;
    let result = if strict_semantics.unwrap_or(false) {
        export_project_workflow_profile_strict(
            &path,
            &preset_id,
            &project_type,
            active_change_id.as_deref(),
            selected_modules.as_deref(),
            disabled_stages.as_deref(),
        )
    } else {
        export_project_workflow_profile(
            &path,
            &preset_id,
            &project_type,
            active_change_id.as_deref(),
            selected_modules.as_deref(),
            disabled_stages.as_deref(),
        )
    };
    result.map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn preview_flow_config_export_cmd(
    state: State<'_, AppState>,
    project_id: String,
    preset_id: String,
    project_type: String,
    selected_modules: Option<Vec<String>>,
    disabled_stages: Option<Vec<String>>,
) -> Result<FlowWritePlan, String> {
    let path = project_path_for_id(&state.db, &project_id)?;
    preview_flow_config_export(
        &path,
        &preset_id,
        &project_type,
        selected_modules.as_deref(),
        disabled_stages.as_deref(),
    )
    .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn export_flow_config_cmd(
    state: State<'_, AppState>,
    project_id: String,
    preset_id: String,
    project_type: String,
    selected_modules: Option<Vec<String>>,
    disabled_stages: Option<Vec<String>>,
    strict_semantics: Option<bool>,
) -> Result<FlowConfig, String> {
    let path = project_path_for_id(&state.db, &project_id)?;
    let result = if strict_semantics.unwrap_or(false) {
        export_flow_config_strict(
            &path,
            &preset_id,
            &project_type,
            selected_modules.as_deref(),
            disabled_stages.as_deref(),
        )
    } else {
        export_flow_config(
            &path,
            &preset_id,
            &project_type,
            selected_modules.as_deref(),
            disabled_stages.as_deref(),
        )
    };
    result.map_err(|e| e.to_string())
}
